{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e3cd5897",
   "metadata": {},
   "source": [
    "# README\n",
    "\n",
    "This notebook implements an analysis pipeline for comparing gene expression\n",
    "levels in non‑switch genomic regions between liver and brain tissues during\n",
    "mouse embryonic development. By focusing on compartment‑stable (non‑switch) \n",
    "regions, we can evaluate whether genes located in the same nuclear compartment \n",
    "exhibit different expression patterns across tissues.\n",
    "\n",
    "## Workflow\n",
    "\n",
    "1. **Data preparation**  \n",
    "    - Load a reference gene location BED file and merge transcripts so that\n",
    "      each gene is represented by its minimal start and maximal end.\n",
    "    - Read a preprocessed single‑cell RNA‑seq AnnData file, filter for high\n",
    "      quality cells.\n",
    "\n",
    "2. **E1 value processing**  \n",
    "    - For each sample/timepoint, read the E1 eigenvector data for liver and\n",
    "      brain.\n",
    "    - Compute the absolute difference between liver and brain E1 values,\n",
    "      identify the top 10 % most divergent bins (switch regions) and the\n",
    "      remaining bins (non‑switch regions).\n",
    "\n",
    "3. **Gene extraction and expression calculation**  \n",
    "    - Map non‑switch bins to genes using genomic coordinates.\n",
    "    - Subset the expression matrix to those genes and compute the mean\n",
    "      expression per gene in liver and brain cells separately.\n",
    "\n",
    "4. **Statistical testing and plotting**  \n",
    "    - Perform a Kolmogorov–Smirnov test on the gene‑wise expression\n",
    "      distributions between tissues.\n",
    "    - Generate joint density plots of log‑transformed mean expression vs.\n",
    "      E1 value, colored by tissue, and save them to a PDF.\n",
    "\n",
    "## Inputs and outputs\n",
    "\n",
    "- **Inputs**:  \n",
    "  - `refgene.bed` gene locations.  \n",
    "  - Eigenvector CSVs (`*_cis_eigs.csv`) for each sample and tissue.  \n",
    "  - Single‑cell expression `Mouse_embryo_all_stage.h5ad`.\n",
    "\n",
    "- **Outputs**:  \n",
    "  - CSV files listing switch ranges per sample.  \n",
    "  - PDF of KDE plots for non‑switch gene expression.  \n",
    "  - Console logs of KS test p‑values.\n",
    "\n",
    "This notebook is intended for reproducible analysis of compartmental gene\n",
    "expression differences and can be adapted to other tissue comparisons or\n",
    "developmental stages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "18b7f60a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "import seaborn as sns\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "\n",
    "e1_dir = '/home/goubo/CRICK/CRICK/spaceA/higashi_v2/higashi/'\n",
    "exp_dir = '/home/goubo/CRICK/scpaRNA/stereo/processeddata/'\n",
    "save_dir = '/home/xuyuetong/CRICK_Data_v3/Paper_Fig/NoSwitchRange_Expression/'\n",
    "geneloc_dir = '/home/goubo/CRICK/CRICK/spaceA/bulkhicres/com_switch/' \\\n",
    "'dchic_E13.5brain_VS_E13.5liver/DifferentialResult/dchic_E13.5brain_VS_E13.5liver/geneEnrichment/'\n",
    "\n",
    "tissue_list = ['Liver', 'Brain']\n",
    "sample_list = ['E11.5L1', 'E11.5L2', 'E12.5L5', 'E12.5L6', 'E13.5C1', 'E13.5C4', 'E13.5C6', 'E14.5F5', 'E14.5F6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fd4413a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_gene_by_range(gene_loc_df, genome_range_df):\n",
    "    gene_list = []\n",
    "    for _, row in genome_range_df.iterrows():\n",
    "        chr_id = row['chrom']\n",
    "        start = row['start']\n",
    "        end = row['end']\n",
    "        matched_genes = gene_loc_df[(gene_loc_df['chrom'] == chr_id) & \n",
    "                                    (gene_loc_df['start'] >= start) & \n",
    "                                    (gene_loc_df['end'] <= end)]\n",
    "        if len(matched_genes) > 0:\n",
    "            gene_list.extend(matched_genes['gene_name'].tolist())\n",
    "    return gene_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06a4015d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      chrom     start       end      gene_name\n",
      "0      chr1   3214482   3671498           Xkr4\n",
      "1      chr1   4119866   4409241            Rp1\n",
      "2      chr1   4490928   4497354          Sox17\n",
      "3      chr1   4773200   4785726         Mrpl15\n",
      "4      chr1   4807823   4846735         Lypla1\n",
      "...     ...       ...       ...            ...\n",
      "25324  chrY  88534803  88578064        Gm20819\n",
      "25325  chrY  89328651  89330983   LOC115489301\n",
      "25326  chrY  89998183  90000531        Gm20879\n",
      "25327  chrY  90751140  90839406  G530011O06Rik\n",
      "25328  chrY  90785442  90816465          Erdr1\n",
      "\n",
      "[25329 rows x 4 columns]\n"
     ]
    }
   ],
   "source": [
    "# Gene locations may have different transcripts, so we take the minimum start and maximum end position for each gene to merge the gene location information.\n",
    "\n",
    "gene_loc_path = '{0}refgene.bed'.format(geneloc_dir)\n",
    "gene_loc_df = pd.read_csv(gene_loc_path, header=None, index_col=None, sep='\\t')\n",
    "gene_loc_df.columns = ['chrom', 'start', 'end', 'gene_name']\n",
    "gene_loc_merge_df = gene_loc_df.groupby(['chrom', 'gene_name']).agg(\n",
    "    start=('start', 'min'),\n",
    "    end=('end', 'max')).reset_index()\n",
    "gene_loc_merge_df = gene_loc_merge_df[['chrom', 'start', 'end', 'gene_name']]\n",
    "gene_loc_merge_df.sort_values(by=['chrom', 'start'], inplace=True, ignore_index=True)\n",
    "print(gene_loc_merge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "36e2a6de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "View of AnnData object with n_obs × n_vars = 478354 × 23761\n",
      "    obs: 'annotation', 'timepoint', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes'\n",
      "    var: 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
      "    uns: 'annotation_colors'\n",
      "    obsm: 'spatial'\n",
      "    layers: 'count'\n"
     ]
    }
   ],
   "source": [
    "# Prepare Expression Data\n",
    "\n",
    "exp_path = '{0}Mouse_embryo_all_stage.h5ad'.format(exp_dir)\n",
    "exp_adata = sc.read_h5ad(exp_path)\n",
    "high_quality = exp_adata.obs[\"total_counts\"] > 2000\n",
    "exp_adata = exp_adata[high_quality, :]\n",
    "print(exp_adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b25d2a2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_path = '{0}Kdeplot_NonSwitch_Expression.pdf'.format(save_dir)\n",
    "\n",
    "with PdfPages(plot_path) as pdf:\n",
    "    for s, sample_id in enumerate(sample_list):\n",
    "\n",
    "        # filter Top10% Diff-E1 range (Liver vs Brain)\n",
    "        e1_liver_path = '{0}{1}/{1}_fasthigashi_leiden_anno_man_{2}.cis_eigs.csv'.format(e1_dir, sample_id, \"Liver\")\n",
    "        e1_liver_df = pd.read_csv(e1_liver_path, header=0, index_col=0, sep=',')\n",
    "        e1_liver_df['start_bin'] = e1_liver_df['start'] // 1_000_000\n",
    "        e1_liver_df['start_bin'] = e1_liver_df['chrom'] + '_' + e1_liver_df['start_bin'].apply(str)\n",
    "        e1_liver_array = e1_liver_df['E1'].values\n",
    "\n",
    "        e1_brain_path = '{0}{1}/{1}_fasthigashi_leiden_anno_man_{2}.cis_eigs.csv'.format(e1_dir, sample_id, \"Brain\")\n",
    "        e1_brain_df = pd.read_csv(e1_brain_path, header=0, index_col=0, sep=',')\n",
    "        e1_brain_df['start_bin'] = e1_brain_df['start'] // 1_000_000\n",
    "        e1_brain_df['start_bin'] = e1_brain_df['chrom'] + '_' + e1_brain_df['start_bin'].apply(str)\n",
    "        e1_brain_array = e1_brain_df['E1'].values\n",
    "\n",
    "        e1_diff = abs(e1_brain_array - e1_liver_array)\n",
    "        e1_diff = np.nan_to_num(e1_diff)\n",
    "        top_count = int(np.ceil(len(e1_diff) * 0.1))\n",
    "        switch_indices = np.argsort(e1_diff)[::-1][:top_count]\n",
    "        noswitch_indices = np.argsort(e1_diff)[::-1][top_count:]\n",
    "        switch_df = e1_liver_df.iloc[switch_indices, :]\n",
    "        noswitch_df = e1_liver_df.iloc[noswitch_indices, :]\n",
    "        switch_df.to_csv('{0}SwitchRange_{1}.csv'.format(save_dir, sample_id), index=False)\n",
    "\n",
    "        # extract gene by noswitch range\n",
    "        noswitch_gene = extract_gene_by_range(gene_loc_merge_df, noswitch_df)\n",
    "        noswitch_gene = [gene for gene in noswitch_gene if gene in exp_adata.var_names]\n",
    "        noswitch_geneloc_df = gene_loc_merge_df[gene_loc_merge_df['gene_name'].isin(noswitch_gene)].copy()\n",
    "        noswitch_geneloc_df['start_bin'] = noswitch_geneloc_df['start'] // 1_000_000\n",
    "        noswitch_geneloc_df['start_bin'] = noswitch_geneloc_df['chrom'] + '_' + noswitch_geneloc_df['start_bin'].apply(str)\n",
    "        \n",
    "\n",
    "        # extract noswitch-gene expression\n",
    "        time = sample_id[:5]\n",
    "        liver_exp_adata = exp_adata[(exp_adata.obs['timepoint'] == time) & (exp_adata.obs['annotation'] == 'Liver')]\n",
    "        liver_noswitchgene_exp = liver_exp_adata[:, noswitch_gene]\n",
    "        if hasattr(liver_noswitchgene_exp.X, 'toarray'):\n",
    "            liver_column_means = np.mean(liver_noswitchgene_exp.X.toarray(), axis=0)\n",
    "        else:\n",
    "            liver_column_means = np.mean(liver_noswitchgene_exp.X, axis=0)\n",
    "        liver_noswitchgene_expmeans = pd.DataFrame({'gene_name':liver_noswitchgene_exp.var_names, 'exp_mean': liver_column_means, 'tissue_id': 'Liver'})\n",
    "        liver_noswitchgene_expmeans = liver_noswitchgene_expmeans.merge(noswitch_geneloc_df, on=['gene_name'], how='left')\n",
    "        bin_e1_liver_dict = dict(e1_liver_df[['start_bin', 'E1']].values)\n",
    "        liver_noswitchgene_expmeans['E1'] = liver_noswitchgene_expmeans['start_bin'].map(bin_e1_liver_dict)\n",
    "\n",
    "        brain_exp_adata = exp_adata[(exp_adata.obs['timepoint'] == time) & (exp_adata.obs['annotation'] == 'Brain')]\n",
    "        brain_noswitchgene_exp = brain_exp_adata[:, noswitch_gene]\n",
    "        if hasattr(brain_noswitchgene_exp.X, 'toarray'):\n",
    "            brain_column_means = np.mean(brain_noswitchgene_exp.X.toarray(), axis=0)\n",
    "        else:\n",
    "            brain_column_means = np.mean(brain_noswitchgene_exp.X, axis=0)\n",
    "        brain_noswitchgene_expmeans = pd.DataFrame({'gene_name':brain_noswitchgene_exp.var_names, 'exp_mean': brain_column_means, 'tissue_id': 'Brain'})\n",
    "        brain_noswitchgene_expmeans = brain_noswitchgene_expmeans.merge(noswitch_geneloc_df, on=['gene_name'], how='left')\n",
    "        bin_e1_brain_dict = dict(e1_brain_df[['start_bin', 'E1']].values)\n",
    "        brain_noswitchgene_expmeans['E1'] = brain_noswitchgene_expmeans['start_bin'].map(bin_e1_brain_dict)\n",
    "\n",
    "        ks_statistic, ks_pvalue = stats.ks_2samp(liver_column_means, brain_column_means)\n",
    "        print(f\"{sample_id}\\tKS test: {ks_pvalue}\")\n",
    "        \n",
    "        noswitchgene_expmeans = pd.concat([liver_noswitchgene_expmeans, brain_noswitchgene_expmeans], ignore_index=True)\n",
    "        noswitchgene_expmeans['log_exp_mean'] = noswitchgene_expmeans['exp_mean'].apply(lambda x: np.log1p(x))\n",
    "        \n",
    "        # plot\n",
    "        plt.figure(figsize=(4, 4))\n",
    "        j = sns.jointplot(data=noswitchgene_expmeans, x=\"log_exp_mean\", y=\"E1\", hue=\"tissue_id\", kind=\"kde\")\n",
    "        plt.title(sample_id, y=0.95)\n",
    "        plt.xlim(-0.15, 1.00)\n",
    "        plt.xlabel('log(Expression Mean)')\n",
    "        plt.tight_layout()\n",
    "        pdf.savefig()\n",
    "        plt.close()\n",
    "\n",
    "       "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "higashi",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
