{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4bc95995",
   "metadata": {},
   "source": [
    "# README\n",
    "\n",
    "This notebook is part of a project to investigate the relationship between\n",
    "compartment eigenvector (E1) values and epigenetic signals in mouse\n",
    "embryonic tissues at the E14.5 stage.  Earlier analyses showed large\n",
    "systematic differences between time points, so the current focus is on\n",
    "within‑sample comparisons of different organs (liver vs. brain) and\n",
    "chromosomes.\n",
    "\n",
    "## Purpose\n",
    "\n",
    "* compute Pearson correlation coefficients between E1 values and\n",
    "    ChIP‑seq signals (H3K27ac, H3K27me3, H3K9me3) for each chromosome and\n",
    "    tissue,\n",
    "* generate line plots of E1, GC content and selected histone marks for a\n",
    "    representative sample (E14.5F5) on a given chromosome (chr1, chr7, …),\n",
    "    and\n",
    "* overlay “switch” regions where compartment status changes.\n",
    "\n",
    "The notebook reads pre‑computed E1 eigenvectors, GC coverage and\n",
    "histone‑mark bed files at 1 Mb resolution, builds a combined `h3k_df`\n",
    "DataFrame, evaluates correlations, and exports publication‑quality PDFs\n",
    "showing the profiles.\n",
    "\n",
    "## Usage\n",
    "\n",
    "1. Adjust the directory paths at the top of the script to point to your\n",
    "     data repositories.\n",
    "2. Modify `sample_list`, `tissue_list`, `h3k_list` and `time` as needed.\n",
    "3. Run each cell in order; the correlation results will print to the\n",
    "     console and the line‑plot PDFs will be saved in `save_dir`.\n",
    "\n",
    "The analysis supports further exploration of epigenetic‑compartment\n",
    "relationships and selection of high‑correlation samples/regions for\n",
    "figures or downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "829e86f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cooler\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "\n",
    "dir_path = '/home/goubo/CRICK/CRICK/spaceA/higashi_v2/higashi/'\n",
    "gc_dir = '/home/xuyuetong/CRICK_Data_v4/Split_Cluster/'\n",
    "h3k_dir = '/home/goubo/CRICK/CRICK/spaceA/onlineData/epi_nature_BingRen_mouseEmbryo/bed_1mb/'\n",
    "switch_dir = '/home/xuyuetong/CRICK_Data_v3/Paper_Fig/NoSwitchRange_Expression/'\n",
    "save_dir = '/home/xuyuetong/CRICK_Data_v3/Paper_Fig/Lineplot_E1value/'\n",
    "\n",
    "time = '14.5'\n",
    "sample_list = ['E14.5F5', 'E14.5F6']\n",
    "tissue_list = ['Liver', 'Brain']\n",
    "h3k_list = ['H3K27ac', 'H3K27me3', 'H3K9me3']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5df360ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      time  chrom  window_start tissue     class     value\n",
      "0     14.5   chr1             0  Liver   H3K27ac  0.000000\n",
      "1     14.5   chr1       1000000  Liver   H3K27ac  0.000000\n",
      "2     14.5   chr1       2000000  Liver   H3K27ac  0.000000\n",
      "3     14.5   chr1       3000000  Liver   H3K27ac  0.000000\n",
      "4     14.5   chr1       4000000  Liver   H3K27ac  4.973663\n",
      "...    ...    ...           ...    ...       ...       ...\n",
      "9887  14.5  chr19      57000000  Brain  H3K27me3  4.451700\n",
      "9888  14.5  chr19      58000000  Brain  H3K27me3  3.724736\n",
      "9889  14.5  chr19      59000000  Brain  H3K27me3  4.225715\n",
      "9890  14.5  chr19      60000000  Brain  H3K27me3  3.900072\n",
      "9891  14.5  chr19      61000000  Brain  H3K27me3  3.999140\n",
      "\n",
      "[9892 rows x 6 columns]\n"
     ]
    }
   ],
   "source": [
    "# prepare h3k data\n",
    "\n",
    "h3k_df = pd.DataFrame()\n",
    "for h3k_i in h3k_list:\n",
    "    for chrom_i in range(1, 20):\n",
    "        chrom_id = 'chr{0}'.format(chrom_i)\n",
    "        for tissue in tissue_list:\n",
    "            if tissue == 'Liver':\n",
    "                h3k_path = '{0}ChIP-seq_{1}_liver_e{2}.1mb.bed'.format(h3k_dir, h3k_i, time)\n",
    "                h3k_liver_df = pd.read_csv(h3k_path, header=0, index_col=None, sep='\\t')\n",
    "                h3k_array = h3k_liver_df[h3k_liver_df['chrom'] == chrom_id]['mean_signal'].values\n",
    "                win_start = h3k_liver_df[h3k_liver_df['chrom'] == chrom_id]['window_start'].values\n",
    "            else:\n",
    "                h3k_fore_path = '{0}ChIP-seq_{1}_forebrain_e{2}.1mb.bed'.format(h3k_dir, h3k_i, time)\n",
    "                h3k_mid_path = '{0}ChIP-seq_{1}_midbrain_e{2}.1mb.bed'.format(h3k_dir, h3k_i, time)\n",
    "                h3k_hind_path = '{0}ChIP-seq_{1}_hindbrain_e{2}.1mb.bed'.format(h3k_dir, h3k_i, time)\n",
    "                h3k_fore_df = pd.read_csv(h3k_fore_path, header=0, index_col=None, sep='\\t')\n",
    "                h3k_fore_sign = h3k_fore_df[h3k_fore_df['chrom'] == chrom_id]['total_signal'].values\n",
    "                h3k_fore_peak = h3k_fore_df[h3k_fore_df['chrom'] == chrom_id]['peak_count'].values\n",
    "                h3k_mid_df = pd.read_csv(h3k_mid_path, header=0, index_col=None, sep='\\t')\n",
    "                h3k_mid_sign = h3k_mid_df[h3k_mid_df['chrom'] == chrom_id]['total_signal'].values\n",
    "                h3k_mid_peak = h3k_mid_df[h3k_mid_df['chrom'] == chrom_id]['peak_count'].values\n",
    "                h3k_hind_df = pd.read_csv(h3k_hind_path, header=0, index_col=None, sep='\\t')\n",
    "                h3k_hind_sign = h3k_hind_df[h3k_hind_df['chrom'] == chrom_id]['total_signal'].values\n",
    "                h3k_hind_peak = h3k_hind_df[h3k_hind_df['chrom'] == chrom_id]['peak_count'].values\n",
    "                h3k_brain_sign = h3k_fore_sign + h3k_mid_sign + h3k_hind_sign\n",
    "                h3k_brain_peak = h3k_fore_peak + h3k_mid_peak + h3k_hind_peak\n",
    "                h3k_array = np.divide(h3k_brain_sign, h3k_brain_peak, out=np.zeros_like(h3k_brain_sign, dtype=float), where=h3k_brain_peak != 0)\n",
    "                win_start = h3k_fore_df[h3k_fore_df['chrom'] == chrom_id]['window_start'].values\n",
    "            h3k_i_df = pd.DataFrame({\n",
    "                'time': time, \n",
    "                'chrom': chrom_id,\n",
    "                'window_start': win_start,\n",
    "                'tissue': tissue, \n",
    "                'class': h3k_i,\n",
    "                'value': h3k_array\n",
    "            })\n",
    "            h3k_df = pd.concat([h3k_df, h3k_i_df], ignore_index=True)\n",
    "\n",
    "print(h3k_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5f60c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the Pearson correlation coefficient between E1 values and epigenetic signals for each chromosome\n",
    "\n",
    "for s, sample_id in enumerate(sample_list):\n",
    "    for chrom_i in range(1, 20):\n",
    "        chrom_id = 'chr{0}'.format(chrom_i)\n",
    "        for t, tissue in enumerate(tissue_list):\n",
    "            e1_path = '{0}{1}/{1}_fasthigashi_leiden_anno_man_{2}.cis_eigs.csv'.format(dir_path, sample_id, tissue)\n",
    "            e1_df = pd.read_csv(e1_path, header=0, index_col=0, sep=',')\n",
    "            e1_chr_df = e1_df[e1_df['chrom'] == chrom_id].copy()\n",
    "            e1_array = e1_chr_df['E1'].values\n",
    "\n",
    "            # prepare h3k data\n",
    "            for h3k_i in h3k_list:\n",
    "                h3k_array = h3k_df[(h3k_df['time'] == time) & \n",
    "                                   (h3k_df['chrom'] == chrom_id) & \n",
    "                                   (h3k_df['tissue'] == tissue) & \n",
    "                                   (h3k_df['class'] == h3k_i)]['value'].values\n",
    "                valid_mask = ~np.isnan(e1_array) & ~np.isnan(h3k_array)\n",
    "                e1_array_clean = e1_array[valid_mask]\n",
    "                h3k_array_clean = h3k_array[valid_mask]\n",
    "                corr_value = np.corrcoef(e1_array_clean, h3k_array_clean)[0, 1]\n",
    "                print(f'sample id: {sample_id}\\t chrom id: {chrom_id}\\t\\t tissue: {tissue}\\t\\t Pearson between E1 and {h3k_i}:{round(corr_value, 5)}')\n",
    "            print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "af37b904",
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_id = 'E14.5F5'\n",
    "chrom_id = 'chr1'\n",
    "\n",
    "plot_path = '{0}Lineplot_E1_GC_H3K_1MB.pdf'.format(save_dir)\n",
    "\n",
    "with PdfPages(plot_path) as pdf:\n",
    "    \n",
    "    plot_row = len(h3k_list)+1+1        # h3k+E1+GC\n",
    "    plot_col = 2\n",
    "    plt.figure(figsize=(6*plot_col, 2*plot_row))\n",
    "\n",
    "    gc_path = '{0}mm10_gc_cov_1MB.tsv'.format(gc_dir)\n",
    "    gc_cov = pd.read_csv(gc_path, header=0, index_col=None, sep='\\t')\n",
    "    gc_cov_chr = gc_cov.loc[gc_cov['chrom'] == chrom_id, 'GC'].values\n",
    "    gc_cov_chr = gc_cov_chr[3: -3]\n",
    "    for i in range(plot_col):\n",
    "        ax_gc = plt.subplot(plot_row, plot_col, i+1)\n",
    "        ax_gc.plot(gc_cov_chr, label='GC')\n",
    "        ax_gc.set_ylabel('GC', fontsize=16)\n",
    "        ax_gc.set_xticks([])\n",
    "        ax_gc.set_title(chrom_id, fontsize=20)\n",
    "        ax_gc.spines['right'].set_visible(False)\n",
    "        ax_gc.spines['top'].set_visible(False)\n",
    "        ax_gc.spines['bottom'].set_visible(False)\n",
    "\n",
    "    for t, tissue in enumerate(tissue_list):\n",
    "        e1_path = '{0}{1}/{1}_fasthigashi_leiden_anno_man_{2}.cis_eigs.csv'.format(dir_path, sample_id, tissue)\n",
    "        e1_df = pd.read_csv(e1_path, header=0, index_col=0, sep=',')\n",
    "        e1_chr_df = e1_df[e1_df['chrom'] == chrom_id].copy()\n",
    "        e1_array = e1_chr_df['E1'].values\n",
    "\n",
    "        plot_start_idx = plot_col+t+1\n",
    "        ax1 = plt.subplot(plot_row, plot_col, plot_start_idx)\n",
    "        ax1.fill_between(range(e1_chr_df.shape[0]), e1_array, 0, where=(e1_array > 0), facecolor='red', alpha=0.5)\n",
    "        ax1.fill_between(range(e1_chr_df.shape[0]), e1_array, 0, where=(e1_array <= 0), facecolor='blue', alpha=0.5)\n",
    "        ax1.plot(e1_array, label='E1', color='black', alpha=1, lw=0.2)\n",
    "        ax1.set_ylabel('E1', fontsize=16)\n",
    "        ax1.set_title('{0} {1}'.format(sample_id, tissue), fontsize=20)\n",
    "        ax1.set_xticks([])\n",
    "        ax1.spines['right'].set_visible(False)\n",
    "        ax1.spines['top'].set_visible(False)\n",
    "        ax1.spines['bottom'].set_visible(False)\n",
    "\n",
    "        # prepare h3k data\n",
    "        for h, h3k_i in enumerate(h3k_list):\n",
    "            h3k_array = h3k_df[(h3k_df['time'] == time) & \n",
    "                               (h3k_df['chrom'] == chrom_id) & \n",
    "                               (h3k_df['tissue'] == tissue) & \n",
    "                               (h3k_df['class'] == h3k_i)]['value'].values\n",
    "\n",
    "            ax_h3k = plt.subplot(plot_row, plot_col, plot_start_idx+(h+1)*plot_col)   \n",
    "            ax_h3k.plot(h3k_array)\n",
    "            ax_h3k.set_ylabel(h3k_i, fontsize=16)\n",
    "            ax_h3k.set_xticks([])\n",
    "            ax_h3k.spines['right'].set_visible(False)\n",
    "            ax_h3k.spines['top'].set_visible(False)\n",
    "            ax_h3k.spines['bottom'].set_visible(False)\n",
    "\n",
    "    plt.subplots_adjust(hspace=0.5, wspace=0.2)\n",
    "    pdf.savefig()\n",
    "    plt.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e3f5b42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete H3K9me3 information, add switch region information\n",
    "\n",
    "sample_id = 'E14.5F5'\n",
    "chrom_id = 'chr7'\n",
    "h3k_list = ['H3K27ac', 'H3K27me3']\n",
    "\n",
    "switch_region = pd.read_csv('{0}SwitchRange_{1}.csv'.format(switch_dir, sample_id), header=0, index_col=None)\n",
    "switch_region = switch_region[(switch_region['chrom'] == chrom_id)]\n",
    "\n",
    "plot_path = '{0}Lineplot_E1_GC_H3K_1MB.switch.pdf'.format(save_dir)\n",
    "\n",
    "with PdfPages(plot_path) as pdf:\n",
    "    \n",
    "    plot_row = 2*len(h3k_list)+2*1+1        # h3k+E1+GC\n",
    "    plot_col = 1\n",
    "    plt.figure(figsize=(6*plot_col, 2*plot_row))\n",
    "\n",
    "    gc_path = '{0}mm10_gc_cov_1MB.tsv'.format(gc_dir)\n",
    "    gc_cov = pd.read_csv(gc_path, header=0, index_col=None, sep='\\t')\n",
    "    gc_cov_chr = gc_cov.loc[gc_cov['chrom'] == chrom_id, 'GC'].values\n",
    "    gc_cov_chr = gc_cov_chr[3: -3]\n",
    "    ax_gc = plt.subplot(plot_row, plot_col, 1)\n",
    "    ax_gc.plot(gc_cov_chr, label='GC')\n",
    "    ax_gc.set_ylabel('GC', fontsize=16)\n",
    "    ax_gc.set_xticks([])\n",
    "    ax_gc.set_title(chrom_id, fontsize=20)\n",
    "    ax_gc.spines['right'].set_visible(False)\n",
    "    ax_gc.spines['top'].set_visible(False)\n",
    "    ax_gc.spines['bottom'].set_visible(False)\n",
    "\n",
    "    for t, tissue in enumerate(tissue_list):\n",
    "        e1_path = '{0}{1}/{1}_fasthigashi_leiden_anno_man_{2}.cis_eigs.csv'.format(dir_path, sample_id, tissue)\n",
    "        e1_df = pd.read_csv(e1_path, header=0, index_col=0, sep=',')\n",
    "        e1_chr_df = e1_df[e1_df['chrom'] == chrom_id].copy()\n",
    "        e1_array = e1_chr_df['E1'].values\n",
    "\n",
    "        plot_start_idx = plot_col+t*3+1\n",
    "        ax1 = plt.subplot(plot_row, plot_col, plot_start_idx)\n",
    "        ax1.fill_between(range(e1_chr_df.shape[0]), e1_array, 0, where=(e1_array > 0), facecolor='red', alpha=0.5)\n",
    "        ax1.fill_between(range(e1_chr_df.shape[0]), e1_array, 0, where=(e1_array <= 0), facecolor='blue', alpha=0.5)\n",
    "        ax1.plot(e1_array, label='E1', color='black', alpha=1, lw=0.2)\n",
    "        ax1.set_ylabel('E1', fontsize=16)\n",
    "        ax1.set_title('{0} {1}'.format(sample_id, tissue), fontsize=20)\n",
    "        ax1.set_xticks([])\n",
    "        ax1.spines['right'].set_visible(False)\n",
    "        ax1.spines['top'].set_visible(False)\n",
    "        ax1.spines['bottom'].set_visible(False)\n",
    "        for switch_i in range(switch_region.shape[0]):\n",
    "            start = switch_region.iloc[switch_i]['start'] // 1_000_000\n",
    "            end = switch_region.iloc[switch_i]['end'] // 1_000_000\n",
    "            ax1.axvspan(start, end, color='gray', alpha=0.3, label='switch')\n",
    "\n",
    "        # prepare h3k data\n",
    "        for h, h3k_i in enumerate(h3k_list):\n",
    "            h3k_array = h3k_df[(h3k_df['time'] == time) & \n",
    "                               (h3k_df['chrom'] == chrom_id) & \n",
    "                               (h3k_df['tissue'] == tissue) & \n",
    "                               (h3k_df['class'] == h3k_i)]['value'].values\n",
    "\n",
    "            ax_h3k = plt.subplot(plot_row, plot_col, plot_start_idx+(h+1))   \n",
    "            ax_h3k.plot(h3k_array)\n",
    "            ax_h3k.set_ylabel(h3k_i, fontsize=16)\n",
    "            ax_h3k.set_xticks([])\n",
    "            ax_h3k.spines['right'].set_visible(False)\n",
    "            ax_h3k.spines['top'].set_visible(False)\n",
    "            ax_h3k.spines['bottom'].set_visible(False)\n",
    "\n",
    "    plt.subplots_adjust(hspace=0.5, wspace=0.2)\n",
    "    pdf.savefig()\n",
    "    plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "higashi",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
